include/Gaffer/ShufflePlug.inl

```diff
@@ -67,70 +60,126 @@ namespace Gaffer
+		const bool dstReplace = plug->replaceDestinationPlug()->getValue();
+
+		// NOTE : Check if the source plug value contains wildcards. The destination
+		//        plug value cannot contain wildcards but may contain substitutions.
+		//        Any source substitutons should have already been done when evaluating
+		//        the source plug. We need to do destination substitutions manually.
+
+		if( ! IECore::StringAlgo::hasWildcards( srcPattern ) )
 		{
-			auto &shuffle = shuffleValues[i++];
-			if( shuffle.enabled && !shuffle.sourcePattern.empty() )
+			// NOTE : No wildcards in source so shuffle is one move.
+
+			const std::string &srcName = srcPattern;
+			const typename T::const_iterator sIt = sourceContainer.find( srcName );
+			if( sIt != sourceContainer.end() )
 			{
-				// note we've disabled substitutions on the destination plug, so we're performing
-				// both the $source substitution and any other context substitutions here.
-				const std::string destination = scope.context()->substitute( plug->destinationPlug()->getValue() );
-				if( !destination.empty() && IECore::StringAlgo::match( sourceData.first, shuffle.sourcePattern ) )
+				Gaffer::Context::EditableScope scope( Gaffer::Context::current() );
+				scope.set<std::string>( g_sourceVariable, &srcName );
+				const std::string &dstPattern = plug->destinationPlug()->getValue();
+				if( ! dstPattern.empty() )
 				{
-					if( shuffle.replaceDestination || sourceContainer.find( destination ) == sourceContainer.end() )
+					const std::string dstName = scope.context()->substitute( dstPattern );
+					if( srcName != dstName )
 					{
-						destinationContainer[destination] = sourceData.second;
-						if( !newDestinations.insert( destination ).second )
+						if( dstReplace || ( destinationContainer.find( dstName ) == destinationContainer.end() ) )
+						{
+							destinationContainer[ dstName ] = sIt->second;
+							names.insert( dstName );
+						}
+
+						if( srcDelete && ( names.find( srcName ) == names.end() ) )
+						{
+							destinationContainer.erase( srcName );
+						}
+					}
+				}
+			}
+		}
+		else
+		{
+			// NOTE : The source contains wildcards so shuffle may be composed of
+			//        multiple moves. Match source pattern against each source name
+			//        and do destination substitutions.
+
+			Gaffer::Context::EditableScope scope( Gaffer::Context::current() );
+
+			NameContainer moveNames;
+			for( typename T::const_iterator
+					sIt    = sourceContainer.begin(),
+					sItEnd = sourceContainer.end(); sIt != sItEnd; ++sIt )
+			{
+				const std::string &srcName = sIt->first;
+				scope.set<std::string>( g_sourceVariable, &srcName );
+				const std::string &dstPattern = plug->destinationPlug()->getValue();
+				if( ! dstPattern.empty() )
+				{
+					const std::string dstName = scope.context()->substitute( dstPattern );
+					if( ( srcName != dstName ) && ! dstName.empty() && IECore::StringAlgo::match( srcName, srcPattern ) )
```

Comment:
Is there any reason we don't test `StringAlgo::match()` first, and avoid evaluating `destination` when there is no match? In the case of an expression driving `destination`, I don't think a user would expect the expression to be invoked with a non-matching `${source}`. I also suspect that `match()` is quicker than evaluating a plug's value, and that the majority of source values won't match.

Reply:
Its the way the previous code worked, so I kept the same logic, I can try it the way you suggest.

Reply:
done